### ISSUE TITLE
Designing out duplicates items in the timeline

### ISSUE THREAD

**Issue by poljar** [matrix-org/matrix-rust-sdk#1533]
It would likely make sense to solve the de-duplication via a design change. We should switch towards using a `HashMap` and a `Vec` for the Timeline:

```rust
struct Timeline {
    items: HashMap<EventId, TimelineItem>,
    item_positions: Vec<EventId>,
}
```

Note that the exact types don't need to be `Vec` and `HashMap` but they should have properties that are equivalent to those two.

This would prevent us from ever inserting items that have the same `EventId` into the timeline. Another nice advantage of this approach is that, if we have an operation that requires us to find the event by its ID, we can find the event in O(1) time instead of O(N) where N is the number of elements in the timeline.

**Comment by Hywan** [comment]
A [`BTreeMap`](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html) looks ideal. And `im` even has a [`OrdMap`](https://docs.rs/im/latest/im/struct.OrdMap.html) (based on b-trees). Is there any arguments in favor of `IndexMap` rather than `OrdMap`?

**Comment by Hywan** [comment]
![Screenshot 2023-03-08 at 09 21 37](https://user-images.githubusercontent.com/946104/223660070-d8c2600e-cf21-4b42-a07f-a57aa8846645.png)

What? That's probably an error from my side, sorry for that…

**Comment by jplatte** [comment]
> Is there any arguments in favor of `IndexMap` rather than `OrdMap`?

Yes, timeline items have a defined order that we need to present to API users.

**Comment by Hywan** [comment]
`IndexMap` and `OrdMap` both maintain the key order, so why are you suggesting `IndexMap` over `OrdMap` (which is native to `im`)?
